File: front_react/src/lib/roles.js
export const ROLE_PATIENT_SCHEDULER = "patient_scheduler";
export const ROLE_REFERRING_PROVIDER = "referring_provider";

// Legacy roles kept so older accounts do not lose access unexpectedly.
export const ROLE_PROVIDER = "provider";
export const ROLE_ADMIN = "admin";

export const ROLE_PATIENT_SCHEDULER_LABEL = "Receiving provider and scheduler";
export const ROLE_REFERRING_PROVIDER_LABEL = "Referring provider";

export const CHAT_ALLOWED_ROLES = [
  ROLE_PATIENT_SCHEDULER,
  ROLE_PROVIDER,
  ROLE_ADMIN,
];

export const REFERRAL_SEARCH_ALLOWED_ROLES = [
  ROLE_PATIENT_SCHEDULER,
];

export const UPLOAD_ALLOWED_ROLES = [
  ROLE_PATIENT_SCHEDULER,
  ROLE_REFERRING_PROVIDER,
  ROLE_PROVIDER,
  ROLE_ADMIN,
];

export const normalizeRole = (role) => {
  if (!role) {
    return null;
  }

  return String(role).trim().toLowerCase().replace(/-/g, "_");
};

export const canAccessChat = (role) =>
  CHAT_ALLOWED_ROLES.includes(normalizeRole(role));

export const canAccessReferralSearch = (role) =>
  REFERRAL_SEARCH_ALLOWED_ROLES.includes(normalizeRole(role));

export const canAccessUpload = (role) =>
  UPLOAD_ALLOWED_ROLES.includes(normalizeRole(role));

export const getRoleDisplayName = (role) => {
  const normalizedRole = normalizeRole(role);

  if (normalizedRole === ROLE_PATIENT_SCHEDULER) {
    return ROLE_PATIENT_SCHEDULER_LABEL;
  }

  if (normalizedRole === ROLE_REFERRING_PROVIDER) {
    return ROLE_REFERRING_PROVIDER_LABEL;
  }

  return role;
};

export const getTopNavItemsForRole = (role) => {
  const normalizedRole = normalizeRole(role);

  if (normalizedRole === ROLE_PATIENT_SCHEDULER) {
    return [
      { to: "/patients", label: "Patients" },
      { to: "/referral-queue", label: "Referral Search" },
      { to: "/specialists-list", label: "Specialists List" },
    ];
  }

  if (normalizedRole === ROLE_REFERRING_PROVIDER) {
    return [{ to: "/external-provider-upload", label: "Referral Upload" }];
  }

  const items = [];

  if (canAccessChat(normalizedRole)) {
    items.push({ to: "/chat", label: "Patient Triage Chat" });
  }

  if (canAccessReferralSearch(normalizedRole)) {
    items.push({ to: "/referral-queue", label: "Referral Search" });
    items.push({ to: "/specialists-list", label: "Specialists List" });
  }

  if (canAccessUpload(normalizedRole)) {
    items.push({ to: "/external-provider-upload", label: "Referral Upload" });
  }

  return items;
};

export const getDefaultRouteForRole = (role) => {
  const normalizedRole = normalizeRole(role);

  if (normalizedRole === ROLE_PATIENT_SCHEDULER) {
    return "/patients";
  }

  if (normalizedRole === ROLE_REFERRING_PROVIDER) {
    return "/external-provider-upload";
  }

  if (canAccessChat(normalizedRole)) {
    return "/chat";
  }

  if (canAccessReferralSearch(normalizedRole)) {
    return "/patients";
  }

  if (canAccessUpload(normalizedRole)) {
    return "/external-provider-upload";
  }

  return "/";
};

export const getPrimaryActionLabelForRole = (role) => {
  const normalizedRole = normalizeRole(role);

  if (normalizedRole === ROLE_PATIENT_SCHEDULER) {
    return "Go to Patients";
  }

  if (normalizedRole === ROLE_REFERRING_PROVIDER) {
    return "Go to Referral Upload";
  }

  if (canAccessChat(normalizedRole)) {
    return "Go to Patient Triage Chat";
  }

  if (canAccessUpload(normalizedRole)) {
    return "Go to Referral Upload";
  }

  return "Go to Eligio AI";
};

File: backend_flask/app/roles.py
ROLE_PATIENT_SCHEDULER = "patient_scheduler"
ROLE_REFERRING_PROVIDER = "referring_provider"

# Legacy roles kept for backwards compatibility with existing users/data.
ROLE_PROVIDER = "provider"
ROLE_ADMIN = "admin"

DEFAULT_REGISTRATION_ROLE = ROLE_PATIENT_SCHEDULER

REGISTRATION_ALLOWED_ROLES = {
    ROLE_PATIENT_SCHEDULER,
    ROLE_REFERRING_PROVIDER,
}

CHAT_ALLOWED_ROLES = {
    ROLE_PATIENT_SCHEDULER,
    ROLE_PROVIDER,
    ROLE_ADMIN,
}

REFERRAL_SEARCH_ALLOWED_ROLES = {
    ROLE_PATIENT_SCHEDULER,
}

UPLOAD_ALLOWED_ROLES = {
    ROLE_PATIENT_SCHEDULER,
    ROLE_REFERRING_PROVIDER,
    ROLE_PROVIDER,
    ROLE_ADMIN,
}

ROLE_ALIASES = {
    "patient scheduler": ROLE_PATIENT_SCHEDULER,
    "patient-scheduler": ROLE_PATIENT_SCHEDULER,
    "receiving provider and scheduler": ROLE_PATIENT_SCHEDULER,
    "receiving_provider_and_scheduler": ROLE_PATIENT_SCHEDULER,
    ROLE_PATIENT_SCHEDULER: ROLE_PATIENT_SCHEDULER,
    "referring provider": ROLE_REFERRING_PROVIDER,
    "referring-provider": ROLE_REFERRING_PROVIDER,
    ROLE_REFERRING_PROVIDER: ROLE_REFERRING_PROVIDER,
    ROLE_PROVIDER: ROLE_PROVIDER,
    ROLE_ADMIN: ROLE_ADMIN,
}


def normalize_role(raw_role, default=DEFAULT_REGISTRATION_ROLE):
    if raw_role is None:
        return default

    normalized = str(raw_role).strip().lower().replace("-", "_")
    return ROLE_ALIASES.get(normalized, normalized)


def can_access_chat(role):
    return normalize_role(role) in CHAT_ALLOWED_ROLES


def can_access_referral_search(role):
    return normalize_role(role) in REFERRAL_SEARCH_ALLOWED_ROLES


def can_access_upload(role):
    return normalize_role(role) in UPLOAD_ALLOWED_ROLES


def can_register_as(role):
    return normalize_role(role) in REGISTRATION_ALLOWED_ROLES
